---
title: "rnaseq_explore_clustering"
author: "Natalie Davidson"
date: "4/12/2021"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

require(data.table)
require(here)
require(ggplot2)
require(dplyr)
require(readr)
require(SummarizedExperiment)
require(ggrepel)
require('biomaRt')
require("VennDiagram")
require("pheatmap")
require(cluster)
require(factoextra)
require(umap)
require(aricode)
require(amap)
require(dbscan)
require(corrr)
require(ggraph)
require(igraph)
require(tripack)

proj_dir = here()
source(file.path(proj_dir, "/analyze_rnaseq/plot_utils.R"))
source(file.path(proj_dir, "/comparison/utils/file_processing_utils.R"))

set.seed(5)

```

## Overview

This document will try multiple clustering methods on the RNA-Seq data to judge how well the 4 annotated clusters are able to be recapitulated.


**Key file locations:**

* RNA-Seq expression: `/hgsc_characterization/data/rna_seq/salmon_quant_processed/salmon_gene_quant.RDS`
* To translate RNA-Seq sample ids to **suid**: `/hgsc_characterization/reference_data/rna_sample_metadata.txt`
* RNA-Seq ground truth subtypes, derived from Nanostring: `/reference_data/RNAseq_sample_selection_subtable_removed.tsv`
* UBC gene set: `/reference_data/UBC_genes.txt`

**Helper Scripts**. helper scripts + methods are listed here: 

* `/hgsc_characterization/comparison/utils/file_processing_utils.R`
    + format_rnaseq_data
    + get_gene_id_map
    
* `/hgsc_characterization/analyze_rnaseq/plot_utils.R`
    + plot_pca
    + plot_umap


## Read in files

```{r read_files, echo=TRUE, warning=FALSE, message=F}

# read in RNASeq files
rnaseq_expr_file = file.path(proj_dir, 
                    "/data/rna_seq/salmon_quant_processed/salmon_gene_quant.RDS")

rnaseq_trans_file = file.path(proj_dir, 
                    "/reference_data/rna_sample_metadata.txt")
rnaseq_dt = format_rnaseq_data(rnaseq_expr_file, rnaseq_trans_file)

# get subtype annotation
subtype_file = file.path(proj_dir, 
                    "/reference_data/RNAseq_sample_selection_subtable_removed.tsv")
subtype_dt = fread(subtype_file)
subtype_dt = unique(subtype_dt[,c("suid", "HGSCsubtype")])

total_dt = merge(subtype_dt, rnaseq_dt, all.y=T)
total_dt = subset(total_dt, !is.na(HGSCsubtype))


# get UBC genes
ubc_file = file.path(proj_dir, "/reference_data/UBC_genes.txt")
ubc_df <- fread(ubc_file)
colnames(ubc_df)[1] = "hgnc_symbol"
gene_ids_df = get_gene_id_map(unique(ubc_df$hgnc_symbol), 
                                filter_type="hgnc_symbol",
                                attributes= c("hgnc_symbol", "ensembl_gene_id"))
ubc_df = merge(as.data.table(gene_ids_df), ubc_df, all.y=T, by="hgnc_symbol")

```

## Process / Normalize Features

```{r process_features, echo=TRUE, warning=FALSE, message=F}

# first process the data using all genes

# scale the gene expression within each sample
total_dt_scaled = total_dt %>%
                        group_by(suid) %>%
                        mutate(scaled_expr = scale(expr))

# now dcast so we can analyze them
total_dt_cast = dcast(total_dt_scaled, suid+HGSCsubtype ~ ensembl_gene_id, value.var="scaled_expr")
row.names(total_dt_cast) = total_dt_cast$suid
total_dt_cast$suid = paste0("Sample_", total_dt_cast$suid)

# now process the data using UBC genes only

# scale the gene expression within each sample
ubc_dt_scaled = subset(total_dt, hgnc_symbol %in% ubc_df$hgnc_symbol) %>%
                        group_by(suid) %>%
                        mutate(scaled_expr = scale(expr))

# now dcast so we can analyze them
ubc_dt_cast = dcast(ubc_dt_scaled, suid+HGSCsubtype ~ ensembl_gene_id, value.var="scaled_expr")
row.names(ubc_dt_cast) = ubc_dt_cast$suid
ubc_dt_cast$suid = paste0("Sample_", ubc_dt_cast$suid)


# only keep top 25% most variable genes otherwise clustering methods take too long
# 25% arbitrarily selected
total_dt_cast_unnorm = dcast(total_dt_scaled, 
                             suid+HGSCsubtype ~ ensembl_gene_id, 
                             value.var="expr")
var_genes = apply(total_dt_cast_unnorm[,3:ncol(total_dt_cast_unnorm)], 2, var)
keep_genes = names(var_genes)[which(var_genes >= quantile(var_genes, 0.75))]
total_dt_cast = total_dt_cast[,c("suid", "HGSCsubtype", keep_genes)]


```

## Cluster Visualization

```{r viz_samples, out.width="50%", echo=FALSE, warning=FALSE, message=F}


## first set up the plotting for all genes
total_num_df = total_dt_cast
color_df = subset(total_num_df, select=c("suid", "HGSCsubtype"))
colnames(color_df) = c("Sample", "Subtype")
total_num_df = t(total_num_df[,grep("ENSG",colnames(total_num_df))])
colnames(total_num_df) = total_dt_cast$suid


tot_clust_df = t(total_num_df)
rownames(tot_clust_df) = make.names(total_dt_cast$HGSCsubtype, unique=T)



## first set up the plotting for UBC genes
ubc_num_df = ubc_dt_cast
ubc_num_df = t(ubc_num_df[,grep("ENSG",colnames(ubc_num_df))])
colnames(ubc_num_df) = ubc_dt_cast$suid

ubc_clust_df = t(ubc_num_df)
rownames(ubc_clust_df) = make.names(ubc_dt_cast$HGSCsubtype, unique=T)


# now plot PCA and UMAP
# we don't scale because the data was already scaled
plot_pca(total_num_df, title="PCA over all genes", color_df, color_id="Subtype", scale=FALSE)
plot_pca(ubc_num_df, title="PCA over UBC genes", color_df, color_id="Subtype", scale=FALSE)

plot_umap(total_num_df, title="UMAP over all genes", color_df, color_id="Subtype", scale=FALSE)
plot_umap(ubc_num_df, title="UMAP over UBC genes", color_df, color_id="Subtype", scale=FALSE)



```

## Key methods for analysis

### K-means method

```{r kmeans_method, echo=TRUE, warning=FALSE, message=F}

#' Helper method to run kmeans
#' over many possibly K's and tracking the
#' nmi and purity for each k.
#' 
#' 
#' @param in_df, dataframe to cluster
#' @param annot_df, annotation df with HGSCsubtype
#' @param cluster_size_keep, k to use in result to return
#' @param cluster_range, range of K
#' @return list, 1: the result for a specified k. 2: cluster performance df
get_kmeans <- function(in_df, annot_df, cluster_size_keep=4, cluster_range=2:5){
    cluster_num_res = NA
    nmi_df = data.frame(k=cluster_range, NMI=NA, purity=NA)
    for(curr_clust_size in cluster_range){
        
        # do kmeans
        km <- Kmeans(in_df, centers = curr_clust_size, nstart = 25, method="spearman")
        cluster_res = km$cluster
        names(cluster_res) = rownames(in_df)
        if(curr_clust_size == cluster_size_keep){
            cluster_num_res = cluster_res
        }
        
        # get nmi metric
        curr_nmi = NMI(km$cluster, annot_df$HGSCsubtype)
        nmi_df$NMI[which(nmi_df$k == curr_clust_size)] = curr_nmi
        
        # get purity metric
        curr_purity = cluster_purity(km$cluster, annot_df$HGSCsubtype)
        nmi_df$purity[which(nmi_df$k == curr_clust_size)] = curr_purity
    }
    
    return(list(cluster_num_res, nmi_df))
    
}


```

### Heirarchical clustering method

```{r hclust, echo=TRUE, warning=FALSE, message=F}

#' Helper method to run hclust, ward.d2
#' 
#' 
#' @param in_df, dataframe to cluster
#' @param annot_df, annotation df with HGSCsubtype
#' @param cluster_range, range of K
#' @param dist_method, distance metric
#' @return list, 1: the result full heirarchical tree. 2: cluster performance df
get_hclust <- function(in_df, annot_df, cluster_range=2:5, dist_method="spearman"){
        
    # do wardD2
    if(dist_method == "spearman"){
        hm <- hclust(as.dist(1-cor(t(in_df), method="spearman")), method="ward.D2")
    }else{
        hm <- hclust(dist(in_df, method=dist_method), method="ward.D2")
    }
    nmi_df = data.frame(k=cluster_range, NMI=NA, purity=NA)
    for(clust_size in cluster_range){
    
        sub_grp <- cutree(hm, k = clust_size)
        
        # get nmi metric
        curr_nmi = NMI(sub_grp, annot_df$HGSCsubtype)
        nmi_df$NMI[which(nmi_df$k == clust_size)] = curr_nmi
        
        # get purity metric
        curr_purity = cluster_purity(sub_grp, annot_df$HGSCsubtype)
        nmi_df$purity[which(nmi_df$k == clust_size)] = curr_purity
    }

    return(list(hm, nmi_df))
    
}


```


### DBScan method

```{r dbscan, echo=TRUE, warning=FALSE, message=F}

#' Helper method to run DBScan
#' 
#' 
#' @param in_df, dataframe to cluster
#' @param annot_df, annotation df with HGSCsubtype
#' @param eps_est, epsilon for determining neighbors
#' @return dataframe, cluster result
get_dbscan <- function(in_df, annot_df, eps_est=0.3){
      
    tot_clust_dist = as.dist(1-cor(t(in_df), method="spearman"))
    
    # do dbscan
    dbs <- dbscan(tot_clust_dist, eps = eps_est, minPts = 4)
    cluster_res = dbs$cluster
    names(cluster_res) = rownames(in_df)
    
    
    # get nmi metric
    curr_nmi = NMI(dbs$cluster, annot_df$HGSCsubtype)
    print(paste0("NMI: ", curr_nmi))
    
    # get purity metric
    curr_purity = cluster_purity(dbs$cluster, annot_df$HGSCsubtype)
    print(paste0("Purity: ", curr_purity))
    
    return(cluster_res)

}


```


### Graph plotting method

```{r plot_graph_method, echo=TRUE, warning=FALSE, message=F}
# heavily using code from 
# https://drsimonj.svbtle.com/how-to-create-correlation-network-plots-with-corrr-and-ggraph
get_graph_cors <- function(in_df, cluster_res, min_corr){
    
    # make correlation df
    tidy_cors <- t(in_df) %>% 
      correlate(method = "spearman") %>% 
      stretch()
    # make this onto a graph, filtering low correlations
    graph_cors <- tidy_cors %>%
        filter(abs(r) > min_corr) %>%
        graph_from_data_frame(directed = FALSE)
    vert_remaining = vertex_attr(graph_cors, "name", index = V(graph_cors))
    
    # add cluster attribute to graph
    graph_cors <- set_vertex_attr(graph_cors, name="cluster", 
                                  index = V(graph_cors), 
                                  cluster_res[vert_remaining])

    return(graph_cors)
}
```

## Cluster Analyses

### K-means all genes

```{r kmeans_all, out.width="50%", echo=FALSE, warning=FALSE, message=F}

res = get_kmeans(tot_clust_df, total_dt_cast, cluster_size_keep=4, cluster_range = 2:5)
cluster4_res = res[[1]]
nmi_df = res[[2]]

# plot clustering metrics
ggplot(nmi_df, aes(x=k, y=NMI)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Normalized mutual Info over different k")

ggplot(nmi_df, aes(x=k, y=purity)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Purity over different k")

fviz_nbclust(tot_clust_df, Kmeans, method = "silhouette",
            diss = Dist(tot_clust_df, method = "spearman")) +
            labs(subtitle = "Silhouette method")

fviz_nbclust(tot_clust_df, Kmeans, method = "wss",
            diss = Dist(tot_clust_df, method = "spearman"))

# plot cluster membership
cluster_res_df = data.frame(cluster4_res, substr(names(cluster4_res), start=1, stop=6))
table(cluster_res_df)



```


### K-means UBC genes

```{r kmeans_ubc, out.width="50%", echo=FALSE, warning=FALSE, message=F}

# now lets do the same on a reduced number of genes
res = get_kmeans(ubc_clust_df, ubc_dt_cast, cluster_size_keep=4, cluster_range = 2:5)
cluster4_res = res[[1]]
nmi_df = res[[2]]

# plot clustering metrics
ggplot(nmi_df, aes(x=k, y=NMI)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Normalized mutual Info over different k")

ggplot(nmi_df, aes(x=k, y=purity)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Purity over different k")

fviz_nbclust(ubc_clust_df, Kmeans, method = "silhouette",
            diss = Dist(ubc_clust_df, method = "spearman")) +
  labs(subtitle = "Silhouette method")

fviz_nbclust(ubc_clust_df, Kmeans, method = "wss",
            diss = Dist(ubc_clust_df, method = "spearman"))

```

```{r kmeans_ubc_graph, out.width="50%", echo=FALSE, warning=FALSE, message=F}

# plot clustering as a graph spearman corr min = 0.75
graph_cors = get_graph_cors(ubc_clust_df, cluster4_res, min_corr=0.75)
ggraph(graph_cors) +
  geom_edge_link(aes(edge_alpha = abs(r), color = r)) +
  guides(edge_alpha = "none", edge_width = "none") +
  scale_edge_colour_gradientn(limits = c(0.9, 1), colors = c("firebrick2", "dodgerblue2")) +
  geom_node_point(size = 5, aes(shape=as.factor(cluster), color=cluster)) +
  geom_node_text(aes(label = name, color=cluster), repel = TRUE) +
  theme_graph() +
  labs(title = "Correlations between sample, UBC, corr > 0.75")

# plot clustering as a graph spearman corr min = 0.9
graph_cors = get_graph_cors(ubc_clust_df, cluster4_res, min_corr=0.9)
ggraph(graph_cors) +
  geom_edge_link(aes(edge_alpha = abs(r), color = r)) +
  guides(edge_alpha = "none", edge_width = "none") +
  scale_edge_colour_gradientn(limits = c(0.9, 1), colors = c("firebrick2", "dodgerblue2")) +
  geom_node_point(size = 5, aes(shape=as.factor(cluster), color=cluster)) +
  geom_node_text(aes(label = name, color=cluster), repel = TRUE) +
  theme_graph() +
  labs(title = "Correlations between sample, UBC, corr > 0.9")

# plot confusion matr
cluster_res_df = data.frame(cluster4_res, substr(names(cluster4_res), start=1, stop=6))
table(cluster_res_df)


```


### Heirarchical clustering all genes, ward.d2, euclidean

```{r all_hclust_euclidean, out.width="50%", echo=FALSE, warning=FALSE, message=F}

res = get_hclust(tot_clust_df, total_dt_cast, cluster_range=2:5, dist_method="euclidean")
hm = res[[1]]
nmi_df = res[[2]]

# plot stats
ggplot(nmi_df, aes(x=k, y=NMI)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Normalized mutual Info over different k")

ggplot(nmi_df, aes(x=k, y=purity)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Purity over different k")

fviz_nbclust(tot_clust_df, hcut, method = "silhouette",
            diss = Dist(tot_clust_df, method = "euclidean")) +
  labs(subtitle = "Silhouette method")

fviz_nbclust(tot_clust_df, hcut, method = "wss",
            diss = Dist(tot_clust_df, method = "euclidean"))

# plot clustering
gg = fviz_dend(hm, k=4) +
            theme_bw() + 
            ggtitle(paste0("K=4, all genes, euclidean"))
print(gg)

# write results
cluster_res <- cutree(hm, k = 4)
cluster_res_df = data.frame(cluster_res, substr(names(cluster_res), start=1, stop=6))
table(cluster_res_df)

```


### Heirarchical clustering UBC genes, ward.d2, euclidean

```{r ubc_hclust_euclidean, out.width="50%", echo=FALSE, warning=FALSE, message=F}


res = get_hclust(ubc_clust_df, ubc_dt_cast, cluster_range=2:5, dist_method="euclidean")
hm = res[[1]]
nmi_df = res[[2]]

# plot cluster stats
ggplot(nmi_df, aes(x=k, y=NMI)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Normalized mutual Info over different k")

ggplot(nmi_df, aes(x=k, y=purity)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Purity over different k")

fviz_nbclust(ubc_clust_df, hcut, method = "silhouette",
            diss = Dist(ubc_clust_df, method = "euclidean")) +
  labs(subtitle = "Silhouette method")

fviz_nbclust(ubc_clust_df, hcut, method = "wss",
            diss = Dist(ubc_clust_df, method = "euclidean"))

# plot clustering
gg = fviz_dend(hm, k=4) +
            theme_bw() + 
            ggtitle(paste0("K=4, UBC genes, euclidean"))
print(gg)

# plot result table
cluster_res <- cutree(hm, k = 4)
cluster_res_df = data.frame(cluster_res, substr(names(cluster_res), start=1, stop=6))
table(cluster_res_df)

```


### Heirarchical clustering all genes, spearman distance ward.d2

```{r all_hclust_spearman, out.width="50%", echo=FALSE, warning=FALSE, message=F}


res = get_hclust(tot_clust_df, total_dt_cast, cluster_range=2:5, dist_method="spearman")
hm = res[[1]]
nmi_df = res[[2]]


# plot stats
ggplot(nmi_df, aes(x=k, y=NMI)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Normalized mutual Info over different k")

ggplot(nmi_df, aes(x=k, y=purity)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Purity over different k")

fviz_nbclust(tot_clust_df, hcut, method = "silhouette",
            diss = Dist(tot_clust_df, method = "spearman")) +
  labs(subtitle = "Silhouette method")

fviz_nbclust(tot_clust_df, hcut, method = "wss",
            diss = Dist(tot_clust_df, method = "spearman"))
# plot clustering
gg = fviz_dend(hm, k=4) +
            theme_bw() + 
            ggtitle(paste0("K=4, all genes, spearman"))
print(gg)

# plot res table
cluster_res <- cutree(hm, k = 4)
cluster_res_df = data.frame(cluster_res, substr(names(cluster_res), start=1, stop=6))
table(cluster_res_df)

```


### Heirarchical clustering UBC genes, spearman, ward.d2

```{r ubc_hclust_spearman, out.width="50%", echo=FALSE, warning=FALSE, message=F}


res = get_hclust(ubc_clust_df, ubc_dt_cast, cluster_range=2:5, dist_method="spearman")
hm = res[[1]]
nmi_df = res[[2]]

# plot stats
ggplot(nmi_df, aes(x=k, y=NMI)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Normalized mutual Info over different k")

ggplot(nmi_df, aes(x=k, y=purity)) +
    geom_point() + geom_line() +
    theme_bw() +  ggtitle("Purity over different k")

fviz_nbclust(ubc_clust_df, hcut, method = "silhouette",
            diss = Dist(ubc_clust_df, method = "spearman")) +
  labs(subtitle = "Silhouette method")

fviz_nbclust(ubc_clust_df, hcut, method = "wss",
            diss = Dist(ubc_clust_df, method = "spearman"))


# plot clustering
gg = fviz_dend(hm, k=4) +
            theme_bw() + 
            ggtitle(paste0("K=4, UBC genes, spearman"))
print(gg)

# write table
cluster_res <- cutree(hm, k = 4)
cluster_res_df = data.frame(cluster_res, substr(names(cluster_res), start=1, stop=6))
table(cluster_res_df)

```



### DBScan all genes

```{r dbscan_all, out.width="50%", fig.align='center', echo=FALSE, warning=FALSE, message=F}


# this plot is used to get the appropriate distance epsilon
# we will use 0.25
tot_clust_dist = as.dist(1-cor(t(tot_clust_df), method="spearman"))
kNNdistplot(tot_clust_dist, k = 4) 

# run clutering
cluster_res = get_dbscan(tot_clust_df, total_dt_cast, eps_est=0.25)


# plot clustering
graph_cors = get_graph_cors(tot_clust_df, cluster_res, min_corr=0.75)
ggraph(graph_cors) +
  geom_edge_link(aes(edge_alpha = abs(r), color = r)) +
  guides(edge_alpha = "none", edge_width = "none") +
  scale_edge_colour_gradientn(limits = c(0.9, 1), colors = c("firebrick2", "dodgerblue2")) +
  geom_node_point(size = 5, aes(shape=as.factor(cluster), color=cluster)) +
  geom_node_text(aes(label = name, color=cluster), repel = TRUE) +
  theme_graph() +
  labs(title = "Correlations between sample, all genes, corr > 0.75")

cluster_res_df = data.frame(cluster_res, substr(names(cluster_res), start=1, stop=6))
table(cluster_res_df)

```

### DBScan UBC genes

```{r dbscan_ubc, out.width="50%", fig.align='center', echo=FALSE, warning=FALSE, message=F}


# this plot is used to get the appropriate distance epsilon
ubc_clust_dist = as.dist(1-cor(t(ubc_clust_df), method="spearman"))
kNNdistplot(ubc_clust_dist, k = 4) 

# run clutering
cluster_res = get_dbscan(ubc_clust_df, ubc_dt_cast, eps_est=0.13)


# plot clustering
graph_cors = get_graph_cors(ubc_clust_df, cluster_res, min_corr=0.9)
ggraph(graph_cors) +
  geom_edge_link(aes(edge_alpha = abs(r), color = r)) +
  guides(edge_alpha = "none", edge_width = "none") +
  scale_edge_colour_gradientn(limits = c(0.9, 1), colors = c("firebrick2", "dodgerblue2")) +
  geom_node_point(size = 5, aes(shape=as.factor(cluster), color=cluster)) +
  geom_node_text(aes(label = name, color=cluster), repel = TRUE) +
  theme_graph() +
  labs(title = "Correlations between sample, UBC, corr > 0.9")

cluster_res_df = data.frame(cluster_res, substr(names(cluster_res), start=1, stop=6))
table(cluster_res_df)

```
